build-release.sh

```diff
@@ -14,14 +14,55 @@ else
   echo "Disabling binary signing as config.sh does not define the required data."
 fi
 
-function sign {
+sign_windows() {
   if [ $can_sign == 0 ]; then
     return
   fi
   ./osslsigncode -pkcs12 ${SIGN_KEYSTORE} -pass "${SIGN_PASSWORD}" -n "${SIGN_NAME}" -i "${SIGN_URL}" -t http://timestamp.comodoca.com -in $1 -out $1-signed
   mv $1-signed $1
 }
 
+sign_macos() {
+  if [ ! -z "${OSX_HOST}" ]; then
+    osx_tmpdir=$(ssh "${OSX_HOST}" "mktemp -d")
+    reldir="$1"
+    binname="$2"
+    appname="$3"
+
+    entitlements=editor.entitlements
+    if [[ "${appname}" == "Godot_mono.app" ]]; then
+      entitlements=editor_mono.entitlements
+    fi
+
+
+    scp "${reldir}/${binname}.zip" "${OSX_HOST}:${osx_tmpdir}"
+    scp "${basedir}/build-macosx/${entitlements}" "${OSX_HOST}:${osx_tmpdir}"
+    ssh "${OSX_HOST}" "
+              cd ${osx_tmpdir} && \
+              unzip ${binname}.zip && \
+              codesign --deep --timestamp \
+                --options=runtime --entitlements ${entitlements} \
+                -s ${OSX_KEY_ID} -v ${appname}/Contents/MacOS/Godot && \
+              zip -r ${binname}_signed.zip ${appname}"
+
+    request_uuid=$(ssh "${OSX_HOST}" "xcrun altool --notarize-app --primary-bundle-id \"${OSX_BUNDLE_ID}\" --username \"${APPLE_ID}\" --password \"${APPLE_ID_PASSWORD}\" --file ${osx_tmpdir}/${binname}_signed.zip")
+    request_uuid=$(echo ${request_uuid} | sed -e 's/.*RequestUUID = //')
+    ssh "${OSX_HOST}" "while xcrun altool --notarization-history 0 -u \"${APPLE_ID}\" -p \"${APPLE_ID_PASSWORD}\" | grep -q ${request_uuid}.*in\ progress; do echo Waiting on Apple signature; sleep 30s; done"
+    if ! ssh "${OSX_HOST}" "xcrun altool --notarization-history 0 -u \"${APPLE_ID}\" -p \"${APPLE_ID_PASSWORD}\" | grep -q ${request_uuid}.*success"; then
+      echo "Signing failed?"
```

Comment:
Signing is working, but notarization is falling, script should get full details.

```suggestion
      notarization_log=$(ssh "${OSX_HOST}" "xcrun altool notarization-info ${request_uuid} -u \"${APPLE_ID}\" -p \"${APPLE_ID_PASSWORD}\"")
      echo ${notarization_log}
```

Reply:
I was just looking into how to implement that, thanks :D

Reply:
Here's the log:
```
{
  "logFormatVersion": 1,
  "jobId": "305dfa5e-274d-4af3-87fc-c3ae100e5aab",
  "status": "Invalid",
  "statusSummary": "Archive contains critical validation errors",
  "statusCode": 4000,
  "archiveFilename": "Godot_v3.2.4-rc_mono_osx.64_signed.zip",
  "uploadDate": "2021-03-02T13:51:00Z",
  "sha256": "de8ae2ff9323953f1f8b72acd9b4fcf8c68ac88bcd9ba747953a5bf68dd6788f",
  "ticketContents": null,
  "issues": [
    {
      "severity": "error",
      "code": null,
      "path": "Godot_v3.2.4-rc_mono_osx.64_signed.zip/Godot_mono.app/Contents/Resources/GodotSharp/Tools/aot-compilers/iphone-arm64/aarch64-apple-darwin-mono-sgen",
      "message": "The binary is not signed.",
      "docUrl": null,
      "architecture": "x86_64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "Godot_v3.2.4-rc_mono_osx.64_signed.zip/Godot_mono.app/Contents/Resources/GodotSharp/Tools/aot-compilers/iphone-arm64/aarch64-apple-darwin-mono-sgen",
      "message": "The signature does not include a secure timestamp.",
      "docUrl": null,
      "architecture": "x86_64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "Godot_v3.2.4-rc_mono_osx.64_signed.zip/Godot_mono.app/Contents/Resources/GodotSharp/Tools/aot-compilers/iphone-arm64/aarch64-apple-darwin-mono-sgen",
      "message": "The executable does not have the hardened runtime enabled.",
      "docUrl": null,
      "architecture": "x86_64"
    }
  ]
}
```

Reply:
We do `codesign --deep ... -v ${appname}/Contents/MacOS/Godot` so maybe it actually doesn't go through the all `.app`. Would it work with `codesign --deep ... -v ${appname}`?

Reply:
Try singing `aarch64-apple-darwin-mono-sgen` with it's own `codesign` command, before signing the app (and do the same for 3 `dylibs` as well), `--deep` flag might be confused by complex folder structure.

Reply:
> ${appname}

Signing with `${appname}` instead of `${appname}/Contents/MacOS/Godot` also skips `aarch64-apple-darwin-mono-sgen`, so I guess signing each part individually is the only option. Since `--deep` is signing `dylib`s correctly, signing `aarch64-apple-darwin-mono-sgen` and then `${appname}` with `-deep` should also work.

Reply:
Signing `aarch64-apple-darwin-mono-sgen` first then `Godot` seems to work!